### PR TITLE
feat: commentary mention support

### PR DESCRIPTION
**Pull request by sshanzel**
## Changes
- Allow mention in the commentary by using the MarkdownInput but having disabled all the features except mention.
- Refactored the Markdown
  - To allow to completely replace the footer.
  - To make preview feature an optional prop.
  - Also fixed some margin on the post button along the way.
- New modal for CreateSharedPost to be used by the Share button.

### Describe what this PR does
- Short and concise, bullet points can help
- Screenshots if applicable can also help

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-1391 #done

### CODE REVIEW

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    return onSubmitPost(e, squad.id, commentary);
+  };
+
+  const [checkUrl] = useDebounce((value: string) => {
+    const links = [link, updatedPreview?.url, updatedPreview?.permalink];
+    if (!isValidHttpUrl(value) || links.some((url) => url === value))
+      return null;
+
+    return getLinkPreview(value);
+  }, 1000);
```
Comment (rebelchris) [inline packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:57]: We are using this 3 times now I see, perhaps we can simply it in a custom hook to keep it aligned if anything ever changes

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    setLink(value);
+    checkUrl(value);
+  };
+
+  return (
+    <Modal kind={Modal.Kind.FlexibleCenter} size={Modal.Size.Medium} {...props}>
+      <Modal.Header title="New post" />
+      <form
+        className="flex flex-col p-3 w-full"
+        action="#"
```
Comment (rebelchris) [inline packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:70]: This feels a bit odd?

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                className="flex-col-reverse m-3 !w-auto"
+                preview={updatedPreview ?? preview}
+                link={link}
+                onLinkChange={onInput}
+              />
+            )
+          }
+        />
+      </form>
+      <Modal.Footer className="typo-caption1">
```
Comment (rebelchris) [inline packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:97]: Why is this footer different from the guidelines?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Missing a little bit of context around this, so only reviewed it very high level.
Perhaps I can look closer after a quick catch up in our 1:1?

File: packages/shared/src/components/modals/common.tsx [hunk 4]
```diff
@@ -64,6 +71,7 @@ export const modals = {
   [LazyModal.ReadingHistory]: ReadingHistoryModal,
   [LazyModal.SquadPromotion]: SquadPromotionModal,
   [LazyModal.LegoReferralCampaign]: LegoReferralCampaignModal,
+  [LazyModal.CreateSharedPost]: CreateSharedPostModal,
```
Comment (capJavert) [inline packages/shared/src/components/modals/common.tsx:74]: Can we also remove `PostToSquadModal`? 

The new modal should replace it right, functionality wise? If yes we need to make sure `SharePostChecklistStep` and `CompanionMenu` are updated to use this new modal.

File: packages/shared/src/components/fields/MarkdownInput/index.tsx [hunk 5]
```diff
@@ -89,28 +116,46 @@ function MarkdownInput({
...
             ref={textareaRef}
-            className="m-4 bg-transparent outline-none typo-body placeholder-theme-label-quaternary"
-            placeholder="Start a discussion, ask a question or write about anything that you believe would benefit the squad. (Optional)"
+            className={classNames(
+              'flex flex-1 bg-transparent outline-none typo-body placeholder-theme-label-quaternary',
+              showUserAvatar ? 'm-3' : 'm-4',
+            )}
             value={input}
-            onClick={() => checkMention()}
+            onClick={() => checkMention && checkMention()}
```
Comment (capJavert) [inline packages/shared/src/components/fields/MarkdownInput/index.tsx:154]: Would prefer `if` here for verbosity and also return type of this expression.

File: packages/shared/src/components/fields/MarkdownInput/index.tsx [hunk 6]
```diff
@@ -72,6 +97,8 @@ function MarkdownInput({
     enabledCommand,
   });
 
+  useImperativeHandle(ref, () => ({ textareaRef, onMentionCommand }));
```
Comment (capJavert) [inline packages/shared/src/components/fields/MarkdownInput/index.tsx:100]: Just to check if we need this or we use this handle on ref anywhere in the code? Can not see it.

**Review by capJavert — CHANGES_REQUESTED** [review]
Few thing that caught my eye, functionality wise I noticed the input does not get cleared after posting.

image

File: packages/shared/src/components/modals/common.tsx (hunk 4 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/common.tsx:74]: That's right. I was waiting for some answer from the design team on empty state that `PostToSquadModal` provides. I believe they have gotten back to it already. I will see its new implementation.

File: packages/shared/src/components/fields/MarkdownInput/index.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/fields/MarkdownInput/index.tsx:100]: We use this on the Footer button which is located outside the editor. I will line the line of code once I stumble upon it.

Comment (sshanzel) [reply packages/shared/src/components/fields/MarkdownInput/index.tsx:100]: Should be from this line:

https://github.com/dailydotdev/apps/pull/1869/files#diff-eba809a0ead9105aaba7c14726f5ed2bc7c5b59485163dd6b6d9f9de7d8fef9bR104

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:72]: We always `preventDefault`, this is actually safer than having no `action` to ensure the page won't reload.

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx (hunk 3 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:99]: Not sure if there is any custom font size, but afair there is no font specified for `ModalFooter`.

The font used here though is for the screenshot below.

Screenshot 2023-06-15 at 6 05 09 PM

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx]: Good point. I have created a hook for this now.

**Comment by sshanzel** [comment]
@capJavert nice catch! Though it is weird that the input is not getting updated with its controlled value. To fix this, I introduced a ref to manually update the value.

@rebelchris have now made the necessary refactors 🚢

File: packages/shared/src/components/modals/post/CreateSharedPostModal.tsx (hunk 3 above)
Comment (rebelchris) [reply packages/shared/src/components/modals/post/CreateSharedPostModal.tsx:93]: Got it, must be a new addition indeed 🤙

**Review by rebelchris — APPROVED** [review]
Cool, nothing from my side 👍

**Comment by sshanzel** [comment]
Have deployed this in preview2 to make it easier to test. Kindly try sharing some post to a Squad.

https://preview2.app.daily.dev

**Comment by sshanzel** [comment]
Merging now. Ante mentioned to proceed without his final review as we already fixed his concerns.

**Pull request merged**